kpi/tests/test_mock_data_exports.py

```diff
@@ -655,3 +665,58 @@ def test_export_with_disabled_questions(self):
         # fails with `KeyError` prior to fix for kobotoolbox/formpack#219
         self.run_csv_export_test(expected_lines, asset=asset)
 
+    def test_anotheruser_can_export_when_submissions_publicly_shared(self):
+        """
+        Running through behaviour described in issue kpi/#2870 where an asset
+        that has been publicly shared and then explicity shared with a user, the
+        user has lower permissions than an anonymous user and is therefore
+        unable to export submission data.
+        """
+        # resetting permissions of `anotheruser` to have no permissions
+        self.asset.remove_perm(self.anotheruser, PERM_PARTIAL_SUBMISSIONS)
+        self.asset.remove_perm(self.anotheruser, PERM_VIEW_ASSET)
+
+        anonymous_user = get_anonymous_user()
+
+        assert self.asset.has_perm(self.anotheruser, PERM_VIEW_ASSET) == False
+        assert PERM_VIEW_ASSET not in self.asset.get_perms(self.anotheruser)
+
+        self.asset.assign_perm(self.anotheruser, PERM_VIEW_ASSET)
```

Comment:
`PERM_CHANGE_ASSET` implies `PERM_VIEW_ASSET`, this line can be (safely) removed if the line just below is kept.

https://github.com/kobotoolbox/kpi/blob/2870-publicly-shared-permissions-fix/kpi/models/asset.py#L609-L617

Reply:
Good to know 👌